contracts/non-fungible-tokens/andromeda-cw721-staking/src/contract.rs

```diff
@@ -0,0 +1,1072 @@
+            operators: None,
+            modules: None,
+            primitive_contract: None,
+        },
+    )
+}
+
+#[entry_point]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::AndrReceive(msg) => {
+            ADOContract::default().execute(deps, env, info, msg, execute)
+        }
+        ExecuteMsg::ReceiveNft(msg) => handle_receive_cw721(deps, env, info, msg),
+        ExecuteMsg::Claim { key } => execute_claim(deps, env, info, key),
+        ExecuteMsg::Unstake { key } => execute_unstake(deps, env, info, key),
+        ExecuteMsg::UpdateAllowedContracts { contracts } => {
+            execute_update_allowed_contracts(deps, info, contracts)
+        }
+        ExecuteMsg::AddAllowedContract { new_contract } => {
+            execute_add_allowed_contract(deps, info, new_contract)
+        }
+        ExecuteMsg::RemoveAllowedContract { old_contract } => {
+            execute_remove_allowed_contract(deps, info, old_contract)
+        }
+        ExecuteMsg::UpdateOwner { address } => {
+            ADOContract::default().execute_update_owner(deps, info, address)
+        }
+        ExecuteMsg::UpdateUnbondingPeriod { new_period } => {
+            execute_update_unbonding_period(deps, info, new_period)
+        }
+    }
+}
+
+fn execute_update_unbonding_period(
+    deps: DepsMut,
+    info: MessageInfo,
+    new_period: u64,
+) -> Result<Response, ContractError> {
+    let contract = ADOContract::default();
+
+    // Only owner or operator can use this function
+    require(
+        contract.is_owner_or_operator(deps.storage, info.sender.as_str())?,
+        ContractError::Unauthorized {},
+    )?;
+
+    // Save new unbonding period
+    UNBONDING_PERIOD.save(deps.storage, &new_period)?;
+
+    Ok(Response::new().add_attribute("action", "updated_unbonding_time"))
+}
+
+fn execute_update_allowed_contracts(
+    deps: DepsMut,
+    info: MessageInfo,
+    contracts: Vec<String>,
+) -> Result<Response, ContractError> {
+    let contract = ADOContract::default();
+
+    // Only owner or operator can use this function
+    require(
+        contract.is_owner_or_operator(deps.storage, info.sender.as_str())?,
+        ContractError::Unauthorized {},
+    )?;
+
+    ALLOWED_CONTRACTS.save(deps.storage, &contracts)?;
+    Ok(Response::new().add_attribute("action", "updated_allowed_contracts"))
+}
+
+fn execute_add_allowed_contract(
+    deps: DepsMut,
+    info: MessageInfo,
+    new_contract: String,
+) -> Result<Response, ContractError> {
+    let contract = ADOContract::default();
+
+    // Only owner or operator can use this function
+    require(
+        contract.is_owner_or_operator(deps.storage, info.sender.as_str())?,
+        ContractError::Unauthorized {},
+    )?;
+    let mut new_contracts = ALLOWED_CONTRACTS.load(deps.storage)?;
+    new_contracts.push(new_contract);
```

Comment:
Need to check for duplicates here I think

Reply:
Would using a map, similar to the address list, make sense as opposed to an array? The `execute_update_allowed_contracts` function may not function though (at least not easily).

Reply:
What added benefits would a map provide compared to using a vector?

Reply:
Faster lookup for inclusion/removal

Reply:
Not expecting the vector to grow large enough that it matters so maybe a vector is fine